core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -1829,7 +1895,52 @@ private[spark] class BlockManager(
     data.dispose()
   }
 
+  /**
+   * Class to handle block manager decommissioning retries
+   * It creates a Thread to retry offloading all RDD cache blocks
+   */
+  private class BlockManagerDecommissionManager(conf: SparkConf) {
+    @volatile private var stopped = false
+    private val blockReplicationThread = new Thread {
+      override def run(): Unit = {
+        while (blockManagerDecommissioning && !stopped) {
+          try {
+            logDebug("Attempting to replicate all cached RDD blocks")
+            decommissionRddCacheBlocks()
+            logInfo("Attempt to replicate all cached blocks done")
+            val sleepInterval = conf.get(
+              config.STORAGE_DECOMMISSION_REPLICATION_REATTEMPT_INTERVAL)
+            Thread.sleep(sleepInterval)
+          } catch {
+            case _: InterruptedException =>
+              // no-op
+            case NonFatal(e) =>
+              logError("Error occurred while trying to " +
+                "replicate cached RDD blocks for block manager decommissioning", e)
+          }
+        }
+      }
+    }
+    blockReplicationThread.setDaemon(true)
+    blockReplicationThread.setName("block-replication-thread")
+
+    def start(): Unit = {
+      logInfo("Starting block replication thread")
+      blockReplicationThread.start()
+    }
+
+    def stop(): Unit = {
+      if (!stopped) {
+        stopped = true
+        logInfo("Stopping block replication thread")
+        blockReplicationThread.interrupt()
+        blockReplicationThread.join()
```

Comment:
My thought is joining the thread here might block if the interrupt didn't work as we want it to, what do you think?

Reply:
@holdenk This entire BlockManagerDecommissionManager#stop() method is called by `decommissionManager.foreach(_.stop())` inside BlockManager#stop(). 

decommissionManager is None by default (as storage decommissioning feature is behind a config and by default disabled). So this code shouldn't event trigger.

Reply:
Sure but when it is turned on (in your test case) this might keep the worker process from exiting when we ask it to stop?

Reply:
Yeah - But all the tests that are failing in jenkins build are not the ones written in this PR. So that means all those tests must be running with storage-decommissioning-flag disabled?

1. https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121988/testReport/
2. https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121917/testReport/

Is the same Spark application going to get used across all these tests? I was assuming that new SparkApp will be created and destroyed for my specific tests (as BlockManagerDecommissionSuite creates new SparkContext as part of test).

Reply:
So it's hanging with remaining alive workers. Try taking this out and see if it makes a difference.

Reply:
> My thought is joining the thread here might block if the interrupt didn't work as we want it to, what do you think?

I think the initial concern is regarding join waiting infinitely as the "stop"  is not percolated deep enough to stop early.... May be it needs to be refactored?